src/device/ping-device.cpp

```diff
@@ -0,0 +1,94 @@
+#include <ping-device.h>
+
+#include <ping-message-common.h>
+#include <ping-message.h>
+#include <ping-time.h>
+
+bool PingDevice::initialize() {
+  return request(CommonId::PROTOCOL_VERSION)
+    && request(CommonId::DEVICE_INFORMATION);
+}
+
+ping_message* PingDevice::read() {
+  char b;
+  int result = _port.read(&b, 1);
+  if (result > 0) {
+    if (_parser.parseByte(static_cast<char>(b)) == PingParser::NEW_MESSAGE) {
+      return &_parser.rxMessage;
+    }
+  }
+  return nullptr;
+}
+
+ping_message* PingDevice::request(uint16_t id, int timeoutMs) {
+  common_general_request msg;
+  msg.set_requested_id(id);
+  writeMessage(msg);
+  ping_message* reply = waitMessage(id, timeoutMs);
+  if (reply && reply->message_id() == CommonId::NACK) {
+    common_nack* m = static_cast<common_nack*>(reply);
+    if (m->nacked_id() != CommonId::GENERAL_REQUEST) {
```

Comment:
should nacked id be different of `id` and not `CommonId::GENERAL_REQUEST` ?

Reply:
when a request is not accepted (device sends a nack message), the nacked_id == GENERAL_REQUEST. It's a short-coming of the protocol we need to look at the errror_message text to know more about which id was requested.